### ISSUE TITLE
Make Pd scheduler block size settable

### ISSUE THREAD

**Issue by Spacechild1** [pure-data/pure-data#1549]
Currently, Pd uses a fixed scheduler block size of 64 samples (`DEFDACBLKSIZE` in `s_stuff.h`). I think it could be settable at runtime instead, so Pd can also run with hardware buffer sizes smaller than 64 samples.

Actually, `m_pd.h` already has `sys_getblksize`. At the moment, it always returns `DEFDACBLKSIZE`, but we could instead set this value via a command line option, e.g. `-schedblocksize`. Then we just have to replace all the relevant occurances of `DEFDACBLKSIZE` (and similar constants like `DEFSENDVS` and `DEFDELVS`) with `sys_getblksize()`.

---

There is also the general issue that Pd won't run if the hardware buffer size is smaller than the scheduler block size. The question is how to deal with this:
a) buffer the input/output appropriately
b) automatically decrease the scheduler block size as necessary

Personally, I prefer a) because I think the scheduler block size should only be settable *once* at startup and never change afterwards.

**Comment by giuliomoro** [comment]
Some externals rely on `s_stuff.h`'s `#define DEFDACBLKSIZE 64`. To avoid breaking all externals that rely on it, we could replace the `define` with a `#define DEFDACBLKSIZE sys_getblksize()`, or `extern int DEFDACBLKSIZE;`. All externals will need to be rebuilt, but (I guess) only those that use it for static memory allocation would require any changes.

I have also witnessed a least one external which internally hardcoded `64` without relying on `s_stuff.h`. Unfortunately I don't know what that was.

Is there a way to detect at runtime whether an external assumes a blocksize of 64 so that a runtime error can be thrown?

**Comment by Spacechild1** [comment]
> Some externals rely on s_stuff.h's #define DEFDACBLKSIZE 64.

This is unfortunate. Externals should always use `sys_getblklsize` (in `m_pd.h`) to get the scheduler block size.

Since we can't maintain binary compatibility (`DEFDACBLKSIZE` is a constant) and we can only "fix" it by recompiling, why not throw an error and force the external author to fix their code?

E.g.:
```
#ifdef PD_INTERNAL
#define DEFDACBLKSIZE 64
#else
#define DEFDACBLKSIZE <some error message mechanism>
#endif
```

I'm wondering, however, which kind of external would need to know Pd's block size in the first place? Do you remember any examples?

---

Another problem is that currently `DEFDACBLKSIZE` has a double function:
1) default hardware buffer size
2) scheduler block size

In the future, we should reserve this constant for 1) and add a seperate constant for 2) (e.g. `DEFSCHEDBLKSIZE`)

**Comment by Spacechild1** [comment]
To add more to the confusion, `s_stuff.h` also has `get_sys_schedblocksize`, which returns `STUFF->st_schedblocksize`.

My question is, what is `sys_getblksize` supposed to return!?

**Comment by giuliomoro** [comment]
> I'm wondering, however, which kind of external would need to know Pd's block size in the first place? Do you remember any examples?

one that I could find a reference to was called `[xcorr~]`, but I am not sure if it was ever published (and I don't have its source code, just a broken dropbox link). It was doing cross-correlation and it may have used a hardcoded the size of some internal buffer. Point is: as long as `DEFDACBLKSIZE` was there in the semi-public header, external developers may have called that instead of `sys_getblksize()` ...

**Comment by Spacechild1** [comment]
> It was doing cross-correlation and it may have used a hardcoded the size of some internal buffer. 

But why? Usually, DSP externals would use the block size obtain in the "dsp" method.

> Point is: as long as DEFDACBLKSIZE was there in the semi-public header, external developers may have called that instead of sys_getblksize() ...

Sure. This just shows again how important it is to *document* an API :-)

**Comment by giuliomoro** [comment]
Downloading [all available sources](https://github.com/pure-data/deken/issues/251#issuecomment-1014083313) from deken and then grepping through the codebase, there is only one case where `DEFDACBLKSIZE` is used
```
iem16[v1.2](Sources)/iem16/del16write~.c:#ifdef DEFDACBLKSIZE
iem16[v1.2](Sources)/iem16/del16write~.c:# define IEM16DEFBLKSIZE DEFDACBLKSIZE
```

but this:
```
grep -RI "\<64\>" * | grep "\.h:\|\.c:" | grep define
```
highlights several cases where an internal block size-related parameter is defined as 64 (including some instances of
```
#define DEFDELVS 64            /* LATER get this from canvas at DSP time */
```
This is even worse than if they were using `DEFDACBLKSIZE` because there is no way to make it break at build time.

 Btw it seems that `sys_getblksize()` seems widely used.

**Comment by sebshader** [comment]
d_delay.c uses ```#define DEFDELVS 64``` too..

**Comment by giuliomoro** [comment]
For completeness, discussion started here: https://lists.puredata.info/pipermail/pd-list/2022-01/130691.html

@Spacechild1  also added
> The objects in "d_global.c" use their own constant DEFSENDVS (for whatever reason) which you would have to change as well

However, `DEFSENDVS` and `DEFDELVS` are internal to Pd (not defined in a header) so externals could only use them through a lot of effort (i.e.: re-defining them locally as done by some of the externals available on Deken)

**Comment by umlaeute** [comment]
> This is even worse than if they were using DEFDACBLKSIZE because there is no way to make it break at build time.

depends. would the externals actually break if Pd's blocksize was changed?
just because there's a `64` somewhere, does not necessarily mean that "this number must be the same as Pd's blocksize, or else bad things will happen".

**Comment by umlaeute** [comment]
> To add more to the confusion, s_stuff.h also has `get_sys_schedblocksize`, which returns `STUFF->st_schedblocksize`.


here, it actually returns a *pointer* to the value, which doesn't make any sense to me...
it's been there since Pd-0.38 (along with similar functions like `get_sys_schedadvance()`), but i haven't found an external yet that actually uses it (though i *only* searched the ones packaged for Debian)

**Comment by giuliomoro** [comment]
>  (though i only searched the ones packaged for Debian)

not even on Deken: only mentioned in `jit_expr` but unused
```
$ grep -RI get_sys_schedblocksize *
jit_expr[v0.1](Sources)/jit_expr/src/pd.build/pd.def:get_sys_schedblocksize
$ grep -RI get_sys_schedadvance *
jit_expr[v0.1](Sources)/jit_expr/src/pd.build/pd.def:get_sys_schedadvance
```

> would the externals actually break if Pd's blocksize was changed?

very legitimate question, I just wanted to draw attention to the fact that there may be externals making assumptions.

Three spot checks:
`ext13[v0.17.1-4](Windows-amd64-32)(Sources)/ext13/sources/catch13~.c` assumes` DEFSENDVS` to be `64` and fails at runtime with `error("sigcatch13 %s: unexpected vector size", x->x_sym->s_name);` (almost good!)

`"zexy[v2.2.8](Sources)/zexy/src/sfrecord.c` gets an undetected buffer overflow in `sfrecord_perform()` if the number of samples in a block passed to the function is larger than the internally-defined `#define DACBLKSIZE 64`. (bad!)

`windowing[v0.2.0](Sources)/pd-windowing/src/lanczos~.c` preallocates a buffer of the size of its internally-defined `#define DEFBLOCKSIZE 64`  in `lanczos_new()`, but reallocates it in `lanczos_perform()` if the actual blocksize is larger. Already this is a good approach, but it would benefit from calling `sys_getblksize()` from `lanczos_new()`.

**Comment by Spacechild1** [comment]
>> To add more to the confusion, s_stuff.h also has get_sys_schedblocksize, which returns STUFF->st_schedblocksize.
> here, it actually returns a pointer to the value, which doesn't make any sense to me...

There are a couple of these functions. To me it looks like they try to emulate global variables...

I guess `get_sys_schedblocksize` could be deprecated and `sys_getblksize` would remain the recommended (public) API function to obtain the scheduler block size.

**Comment by umlaeute** [comment]
> gets an undetected buffer overflow in sfrecord_perform() [...] (bad!)

indeed. but unrelated to the scheduler blocksize. it's just as bad when using `[block~ 4096]`.
-> https://git.iem.at/pd/zexy/-/issues/10

> but reallocates it in lanczos_perform() if the actual blocksize is larger. Already this is a good approach, but it would benefit from calling sys_getblksize() from lanczos_new() to avoid reallocation in the audio thread.

no. it should resize in the `lanczos_dsp()` instead: this is where the actual per-canvas blocksize is determined (not in the constructor), and it's (conceptually) "outside the audio thread".

**Comment by Spacechild1** [comment]
> no. it should resize in the lanczos_dsp() instead: this is where the actual per-canvas blocksize is determined (not in the constructor), and it's (conceptually) "outside the audio thread".

I guess they want to use a single buffer for all objects. Actually, we do the same in `mayer_dofft`, see `ooura_init`.

Of course, it would be better if functions like `ooura_init` were called in the "dsp" method, to avoid resizing the buffer in the perform routine. However, since the buffers are threadlocal (at least if `PDINSTANCE` is defined), this only works if both methods are called on the same thread. This is usually the case, but there is no guarantee. The actual solution would be to make the FFT buffer really per instance, e.g. store it in `_instancestuff`.

**Comment by umlaeute** [comment]
>>    Some externals rely on s_stuff.h's #define DEFDACBLKSIZE 64.
>
> This is unfortunate. 

on re-considering i think it's not really as bad.

afaict `DEFDACBLKSIZE` really means **def**ault **bl**oc**ksize** to transfer samples to the **dac**.
i read this as a *default* blocksize - the actual blocksize could of course be a  non-default value.

so all code is free to use this value to get a sane default (and 64 is very much sane) but it should make sure to check for different blocksizes as well.

in practice, I agree with @Spacechild1 that there are probably no externals out there that have any use for the dac transfer size.
but that also means, that `sys_getblksize()` is irrelevant for these externals: the per-canvas blocksize is independent of the system blocksize and typically *that* is the value a object is interested in.

**Comment by umlaeute** [comment]
> I guess they want to use a shared buffer for all objects

looking at the [code](https://sources.debian.org/src/pd-windowing/0.1-4/lanczos%7E.c/), i don't think this is true.
it seems to be straight-forward.

**Comment by Spacechild1** [comment]
> looking at the code, i don't think this is true.
it seems to be straight-forward.

You're right. The external is just doing it wrong :-)

**Comment by Spacechild1** [comment]
I agree with @umlaeute: `DEFDACBLKSIZE` can be used as some kind of default value, but any external that *relies*  on this constant to be 64 samples is broken.

> afaict DEFDACBLKSIZE really means default blocksize to transfer samples to the dac.

At the same time, it is also the scheduler block size. In theory, these two values *could* be different, but it wouldn't make much sense.

Fun fact: `_instancestuff` has both `st_schedblocksize` and `st_blocksize`. They are both initialized to the same value (`DEFDACBLKSIZE`), but `st_blocksize` isn't used anywhere.

---

It gets even more confusing: `DEFDACBLKSIZE` is used both as the default scheduler block size *and* the default hardware block size. On the other hand, there are 2 default sample rates: `DEFDACSAMPLERATE` and `DEFAULTSRATE`. The former is practically overwritten by the latter.

I think this whole thing really needs to be cleaned up.

`DEFDACBLKSIZE` should be deprecated and replaced with two distinct constants, e.g.:

a) `DEFAULTBLKSIZE` (= default value of `a_blocksize` in `t_audiosettings`, similar to `DEFAULTSR` and `DEFAULTADVANCE`)
b) `DEFSCHEDBKSIZE` (= default scheduler block size)

`DEFDACSAMPLERATE` could be deprecated in favor of `DEFAULTSR`.
